br/pkg/streamhelper/collector.go

```diff
@@ -0,0 +1,284 @@
+	c.onSuccess = hook
+}
+
+func (c *storeCollector) begin(ctx context.Context) {
+	err := c.recvLoop(ctx)
+	if err != nil {
+		log.Warn("collector loop meet error", logutil.ShortError(err))
+		c.reportErr(err)
+	}
+	close(c.doneMessenger)
+}
+
+func (c *storeCollector) recvLoop(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case r, ok := <-c.input:
+			if !ok {
+				return c.flush(ctx)
+			}
+
+			if r.Leader.StoreId != c.storeID {
+				log.Warn("trying to request to store which isn't the leader of region.",
+					zap.Uint64("region", r.Region.Id),
+					zap.Uint64("target-store", c.storeID),
+					zap.Uint64("leader", r.Leader.StoreId),
+				)
+			}
+			c.appendRegionMap(r)
+			c.currentRequest.Regions = append(c.currentRequest.Regions, &logbackup.RegionIdentity{
+				Id:           r.Region.GetId(),
+				EpochVersion: r.Region.GetRegionEpoch().GetVersion(),
+			})
+			if len(c.currentRequest.Regions) > c.batchSize {
+				err := c.flush(ctx)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+}
+
+func (c *storeCollector) appendRegionMap(r RegionWithLeader) {
+	c.regionMap[r.Region.GetId()] = kv.KeyRange{StartKey: r.Region.StartKey, EndKey: r.Region.EndKey}
+}
+
+type StoreCheckpoints struct {
+	HasCheckpoint    bool
+	Checkpoint       uint64
+	FailureSubranges []kv.KeyRange
+}
+
+func (s *StoreCheckpoints) merge(other StoreCheckpoints) {
+	if other.HasCheckpoint && (other.Checkpoint < s.Checkpoint || !s.HasCheckpoint) {
+		s.Checkpoint = other.Checkpoint
+		s.HasCheckpoint = true
+	}
+	s.FailureSubranges = append(s.FailureSubranges, other.FailureSubranges...)
+}
+
+func (s *StoreCheckpoints) String() string {
+	sb := new(strings.Builder)
+	sb.WriteString("StoreCheckpoints:")
+	if s.HasCheckpoint {
+		sb.WriteString(strconv.Itoa(int(s.Checkpoint)))
+	} else {
+		sb.WriteString("none")
+	}
+	fmt.Fprintf(sb, ":(remaining %d ranges)", len(s.FailureSubranges))
+	return sb.String()
+}
+
+func (c *storeCollector) spawn(ctx context.Context) func() (StoreCheckpoints, error) {
+	go c.begin(ctx)
+	return func() (StoreCheckpoints, error) {
+		close(c.input)
+		<-c.doneMessenger
+		if err := c.Err(); err != nil {
+			return StoreCheckpoints{}, err
+		}
+		sc := StoreCheckpoints{
+			HasCheckpoint:    c.checkpoint != 0,
+			Checkpoint:       c.checkpoint,
+			FailureSubranges: c.inconsistent,
+		}
+		return sc, nil
+	}
+}
+
+func (c *storeCollector) flush(ctx context.Context) error {
+	log.Debug("sending batch", zap.Int("size", len(c.currentRequest.Regions)), zap.Uint64("store", c.storeID))
+	cli, err := c.service.GetLogBackupClient(ctx, c.storeID)
+	if err != nil {
+		return err
+	}
+	cps, err := cli.GetLastFlushTSOfRegion(ctx, &c.currentRequest)
+	if err != nil {
+		return err
+	}
+	metrics.GetCheckpointBatchSize.WithLabelValues("checkpoint").Observe(float64(len(c.currentRequest.GetRegions())))
+	c.currentRequest = logbackup.GetLastFlushTSOfRegionRequest{}
+	for _, checkpoint := range cps.Checkpoints {
```

Comment:
Can it remove region KeyRange information from regionMap to safe memory if it succeeds get checkpoints for a specify region ?

Reply:
We should keep the key range, this is fundamental for the consistency of checkpoint; if the key ranges weren't recorded, we must scan the full key range of the task for getting a safe checkpoint.